Docker/Dockerfile

```diff
@@ -21,12 +23,50 @@ RUN apt-get update \
     clang \
     cmake \
     && rm -rf /var/lib/apt/lists/*
-
 # Install the fmt
 RUN rustup component add rustfmt
+RUN echo "Will build image with mode: ${BUILD_MODE}"
 
-# Set the workdir to the build folder
-WORKDIR "$WORKDIR/build/narwhal"
+#################################################################
+# Stage 1: Planning
+#################################################################
+FROM chef as planner
+ARG BUILD_MODE
+
+# Plan out the 3rd-party dependencies that need to be built.
+#
+# This is done by:
+#   1. Copy in Cargo.toml, Cargo.lock, and the workspace-hack crate
+#   2. Removing all workspace crates, other than the workpsace-hack
+#      crate, from the workspace Cargo.toml file.
+#   3. Update the lockfile in order to reflect the changes to the
+#      root Cargo.toml file.
+COPY Cargo.toml Cargo.lock ./
+COPY workspace-hack workspace-hack
+RUN sed -i '/workspace-hack/b; //d' Cargo.toml \
+    && cargo metadata -q >/dev/null
+
+#################################################################
+# Stage 2 : Caching
+#################################################################
+# Build and cache all dependencies.
+#
+# In a fresh layer, copy in the "plan" generated by the planner
+# and run `cargo build` in order to create a caching Docker layer
+# with all dependencies built.
+FROM chef AS builder 
+ARG BUILD_MODE
+
+COPY --from=planner /narwhal/Cargo.toml Cargo.toml
+COPY --from=planner /narwhal/Cargo.lock Cargo.lock
+COPY --from=planner /narwhal/workspace-hack workspace-hack
+
```

Comment:
You're missing a step here to actually build and cache the dependencies.

Reply:
eg:
```
RUN cargo build --${BUILD_MODE}
```

Reply:
Er, yes, of course, it was supposed to be part of this commit.